File: android/app/src/main/java/dev/disobey/mango/WordPieceTokenizer.kt
package dev.disobey.mango

import com.google.gson.Gson

data class TokenizerOutput(
    val inputIds: LongArray,
    val attentionMask: LongArray,
    val tokenTypeIds: LongArray
)

class WordPieceTokenizer(tokenizerJson: String) {
    private val vocab: Map<String, Int>
    private val unkTokenId = 100  // [UNK]
    private val clsTokenId = 101  // [CLS]
    private val sepTokenId = 102  // [SEP]

    init {
        // Parse tokenizer.json: model.vocab is Map<String, Int>
        data class TokenizerModel(val vocab: Map<String, Int>)
        data class TokenizerJson(val model: TokenizerModel)
        val parsed = Gson().fromJson(tokenizerJson, TokenizerJson::class.java)
        vocab = parsed.model.vocab
        require(vocab.isNotEmpty()) { "Vocabulary is empty" }
    }

    fun tokenize(text: String, maxLength: Int = 256): TokenizerOutput {
        // 1. Lowercase (uncased model)
        val lowered = text.lowercase()

        // 2. Basic tokenization: split on whitespace and punctuation.
        //    Insert spaces around punctuation/symbol chars, then split on whitespace.
        val cleaned = buildString {
            for (ch in lowered) {
                when {
                    ch.isLetterOrDigit() || ch == '\'' -> append(ch)
                    ch.isWhitespace() -> append(' ')
                    else -> append(" $ch ")  // punctuation and other symbols
                }
            }
        }
        val words = cleaned.split(" ").filter { it.isNotEmpty() }

        // 3. WordPiece: greedy longest-match for each word
        val tokenIds = mutableListOf<Int>()
        val maxTokens = maxLength - 2  // Reserve for [CLS] and [SEP]

        for (word in words) {
            if (tokenIds.size >= maxTokens) break
            val wordTokens = wordPiece(word)
            for (t in wordTokens) {
                if (tokenIds.size >= maxTokens) break
                tokenIds.add(t)
            }
        }

        // 4. Assemble output with special tokens: [CLS] ... tokens ... [SEP] [PAD]...
        val inputIds = LongArray(maxLength)        // default 0L = [PAD]
        val attentionMask = LongArray(maxLength)   // default 0L = padding position
        val tokenTypeIds = LongArray(maxLength)    // all zeros (single-sentence input)

        // [CLS] at position 0
        inputIds[0] = clsTokenId.toLong()
        attentionMask[0] = 1L

        // Token positions 1..tokenIds.size
        for (i in tokenIds.indices) {
            inputIds[i + 1] = tokenIds[i].toLong()
            attentionMask[i + 1] = 1L
        }

        // [SEP] after tokens
        val sepPos = tokenIds.size + 1
        inputIds[sepPos] = sepTokenId.toLong()
        attentionMask[sepPos] = 1L

        // Remaining positions left as 0 (PAD token id, zero attention mask)

        return TokenizerOutput(inputIds, attentionMask, tokenTypeIds)
    }

    // Greedy longest-match-first WordPiece sub-word tokenizer.
    // Matches the longest vocabulary entry starting at each position.
    // Uses "##" prefix for continuation sub-words (non-start positions).
    private fun wordPiece(word: String): List<Int> {
        val tokens = mutableListOf<Int>()
        var start = 0
        while (start < word.length) {
            var end = word.length
            var found = false
            while (start < end) {
                val substr = if (start == 0) {
                    word.substring(start, end)
                } else {
                    "##" + word.substring(start, end)
                }
                val id = vocab[substr]
                if (id != null) {
                    tokens.add(id)
                    start = end
                    found = true
                    break
                }
                end--
            }
            if (!found) {
                // Character not in vocab at all; emit [UNK] and advance one character
                tokens.add(unkTokenId)
                start++
            }
        }
        return tokens
    }
}

File: ios/Mango/Mango/WordPieceTokenizer.swift
import Foundation

// MARK: - TokenizerOutput

struct TokenizerOutput {
    let inputIds: [Int64]
    let attentionMask: [Int64]
    let tokenTypeIds: [Int64]
}

// MARK: - EmbeddingError

enum EmbeddingError: Error {
    case modelNotFound
    case tokenizerNotFound
    case vocabEmpty
    case sessionCreationFailed(String)
}

// MARK: - WordPieceTokenizer

/// WordPiece tokenizer for all-MiniLM-L6-v2 (uncased BERT variant).
///
/// Loads vocabulary from tokenizer.json (HuggingFace tokenizers format).
/// Implements greedy longest-match-first WordPiece segmentation.
/// Produces input_ids, attention_mask, and token_type_ids tensors
/// suitable for ONNX Runtime inference.
final class WordPieceTokenizer {
    private let vocab: [String: Int]
    // Special token IDs for BERT uncased vocabulary
    private let unkTokenId: Int = 100   // [UNK]
    private let clsTokenId: Int = 101   // [CLS]
    private let sepTokenId: Int = 102   // [SEP]
    private let padTokenId: Int = 0     // [PAD]

    /// Initialize from tokenizer.json in HuggingFace tokenizers format.
    /// Expected structure: { "model": { "vocab": { "[PAD]": 0, ... } } }
    init(tokenizerURL: URL) throws {
        let data = try Data(contentsOf: tokenizerURL)

        struct Model: Codable {
            let vocab: [String: Int]
        }
        struct TokenizerJSON: Codable {
            let model: Model
        }

        let decoded = try JSONDecoder().decode(TokenizerJSON.self, from: data)
        self.vocab = decoded.model.vocab
        guard !vocab.isEmpty else { throw EmbeddingError.vocabEmpty }
    }

    /// Tokenize a single text string.
    ///
    /// - Parameters:
    ///   - text: Input text to tokenize.
    ///   - maxLength: Total sequence length including [CLS] and [SEP] tokens. Defaults to 256.
    /// - Returns: TokenizerOutput with inputIds, attentionMask, tokenTypeIds all of length maxLength.
    func tokenize(_ text: String, maxLength: Int = 256) -> TokenizerOutput {
        // Step 1: Lowercase (uncased model requires lowercase normalization)
        let lowered = text.lowercased()

        // Step 2: Basic tokenization — insert spaces around punctuation and symbols,
        // then split on whitespace to produce word tokens.
        var cleaned = ""
        cleaned.reserveCapacity(lowered.count * 2)
        for char in lowered {
            if char.isPunctuation || char.isSymbol {
                cleaned += " \(char) "
            } else {
                cleaned.append(char)
            }
        }
        let words = cleaned.split(separator: " ").map(String.init).filter { !$0.isEmpty }

        // Step 3: WordPiece segmentation — greedy longest-match-first for each word.
        // Reserve 2 positions for [CLS] and [SEP] special tokens.
        var tokenIds: [Int] = []
        let maxTokens = maxLength - 2

        for word in words {
            guard tokenIds.count < maxTokens else { break }
            let wordTokens = wordPiece(word)
            for t in wordTokens {
                guard tokenIds.count < maxTokens else { break }
                tokenIds.append(t)
            }
        }

        // Step 4: Prepend [CLS] (101) and append [SEP] (102)
        var inputIds: [Int64] = [Int64(clsTokenId)]
        inputIds.append(contentsOf: tokenIds.map { Int64($0) })
        inputIds.append(Int64(sepTokenId))

        // Step 5: Attention mask — 1 for real tokens, 0 for padding
        let realLen = inputIds.count
        var attentionMask = [Int64](repeating: 1, count: realLen)

        // Step 6: Pad to maxLength with [PAD] (0)
        while inputIds.count < maxLength {
            inputIds.append(Int64(padTokenId))
            attentionMask.append(0)
        }

        // Step 7: token_type_ids — all zeros for single-sentence input
        let tokenTypeIds = [Int64](repeating: 0, count: maxLength)

        return TokenizerOutput(
            inputIds: inputIds,
            attentionMask: attentionMask,
            tokenTypeIds: tokenTypeIds
        )
    }

    // MARK: - WordPiece Algorithm

    /// Greedy longest-match-first WordPiece segmentation.
    ///
    /// For each word, tries to match the longest vocabulary prefix.
    /// Continuation subwords are prefixed with "##" (BERT convention).
    /// Returns [unkTokenId] for any character that cannot be matched.
    private func wordPiece(_ word: String) -> [Int] {
        var tokens: [Int] = []
        var start = word.startIndex

        while start < word.endIndex {
            var end = word.endIndex
            var found = false

            // Try longest match first, shrinking end index until a vocab entry is found
            while start < end {
                let substr: String
                if start == word.startIndex {
                    // First subword: no prefix
                    substr = String(word[start..<end])
                } else {
                    // Continuation subword: "##" prefix per BERT WordPiece convention
                    substr = "##" + String(word[start..<end])
                }

                if let id = vocab[substr] {
                    tokens.append(id)
                    start = end
                    found = true
                    break
                }

                // Shrink match window by one character
                end = word.index(before: end)
            }

            if !found {
                // Cannot match any prefix — emit [UNK] and advance one character
                tokens.append(unkTokenId)
                start = word.index(after: start)
            }
        }

        return tokens
    }
}
